text/0000-natvis.md

```diff
@@ -0,0 +1,495 @@
+`do_some_work()`. (Perhaps some code path has triggered a panic.) Carol wants
+to look at the state of the app, inside the `find_c_defines` function,
+and she specifically wants to see what the state of `captures` is. So she
+selects the `find_c_defines` call frame and looks at the local variables
+window.
+
+Unfortunately, the debugger's view of the `captures` variable does not give
+her any useful information at all. It shows only something like:
+
+```text
+> Variables
+  > captures: {...}
+    > text: "...the entire input text..."
+    > locs: {...}
+      > __0: (4) vec![None, None, None, None]
+      > named_groups: (refs:2) size=0, capacity=1
+        > [raw]: alloc::sync::Arc<std::collections::hash::map::HashMap<...>>
+          > ptr: {pointer:0xNNNNNNNN}
+            > pointer: {...}
+              > strong: {...}
+              > weak: {...}
+              > data: size=0, capacity=1
+                > base: {...}
+                  > hash_builder: {...}
+        ...
+```
+
+The debugger shows the structure of the data, not its meaning. It is not very
+useful for Carol. Even the implementor of `regex` would have a hard time knowing
+how to decode this. In reality, when trying to understand the state of the captures
+variable there are several methods defined for a `Captures` type that paint
+the actual picture in terms of the information a Rust developer would like to
+extract from this variable. In order to meaningfully understand what the `Captures`
+type is truly trying to tell us, it would be very helpful to visualize this data
+differently in the debugger.
+
+What we _want_ is something like this:
+
+```text
+> Variables:
+  > captures: {...}
+    > 1: "SOME_CONSTANT"
+    > 2: "42"
+    > 3: "// some developer comment"
+```
+
+This RFC will describe how to support adding Natvis visualizations which is supported by:
+
+* The Windows Debugger (WinDbg)
+* Visual Studio Debugger.
+
+It should be easy for Rust developers to add debugger visualizations to their
+crates.
+
+## Supporting Natvis
+
+This section describes how Microsoft's Natvis is supported in Rust.
+
+To use Natvis, developers write XML documents that describe how debugger types
+should be displayed using the natvis schema. (See: https://docs.microsoft.com/en-us/visualstudio/debugger/create-custom-views-of-native-objects?view=vs-2019)
+The Natvis files provide patterns, which match type names, and for matching
+types, a description of how to display those types. This allows for some
+limited support for generic types.
+
+Rust developers can add one or more `.natvis` files to their crate. The
+`Cargo.toml` file specifies the path of these Natvis files via a new
+manifest key. Cargo would then pass the set of `.natvis` files that were
+specified in `Cargo.toml` and pass it to the rustc invocation for the crate
+via a new `-C` flag and enabling unstable options in Rust via the
+`-Z unstable-options`. See the below example for how rustc would embed these
```

Comment:
The main reason this RFC requires rustc support is to ensure that any Natvis file specified for a crate would also be stored in the rmeta for that particular crate. I do not believe `link-args` are stored in the metadata today and they are also marked as an untracked option.

Reply:
I see. This is explained in more detail just above "Drawbacks". Sorry for missing that!

I'm not sure what embedding the *paths* to the natvis files in rmeta gives us though. Since the rmetas only contain the paths, in any situation where you're distributing rlibs/rmetas you'll still need to ensure the paths point to the correct files. For example, `libstd` and `libcore` will still need special handling so that the natvis files are included with the Rust distribution artifacts. 

If the natvis files were embedded directly into rmeta, then that would resolve that issue (although I suppose we will need to extract the natvis files from the rmeta before we can pass them to `link.exe`) and the rmeta/rlib would contain everything needed by downstream dependencies. 

Alternatively, I could also imagine that Cargo would collect all the natvis files in dependencies and pass them to the final `rustc` invocation. For example, if we have a crate `binary` that depends on a crate `library`, then `cargo build` would

```bash
# compile the library dependency
rustc --crate-name library library\src\lib.rs --crate-type lib

# compile the binary, passing in all natvis files from the crate graph
rustc --crate-name binary binary\src\main.rs --crate-type bin --extern library=path/to/liblibrary.rlib --natvis library\a.natvis --natvis library\b.natvis --natvis binary\c.natvis
```

Which, if we did that, then we don't really need `--natvis` at all, we can do `-Clink-arg` as I was suggesting above.

Reply:
> If the natvis files were embedded directly into rmeta, then that would resolve that issue (although I suppose we will need to extract the natvis files from the rmeta before we can pass them to link.exe) and the rmeta/rlib would contain everything needed by downstream dependencies.

I believe that this is the correct approach here. Serializing the contents of each natvis file and embed that directly in the rmeta. This way rlibs/rmetas can be distributed and still contain the natvis for each crate. This will now add the additional step of extracting the natvis from the rmeta to ensure they can be passed into the linker.

I'll update the RFC to add this point.